### PR TITLE
feat: :sparkles: init nerdctl support

### PR DESCRIPTION
**Pull request by zwpaper**
Signed-off-by: Wei Zhang <kweizh@gmail.com>

1. upstream buildkit support for nerdctl https://github.com/moby/buildkit/pull/3453
2. use `os/exec` for nerdctl support
3. working on encapsulating docker and nerdctl inside `pkg/driver` https://github.com/zwpaper/envd/blob/5d7660a7727db414d304e431bd1f46384b1d5546/pkg/driver/client.go

fix https://github.com/tensorchord/envd/issues/576

### CODE REVIEW

**Comment by gaocegege** [comment]
Thanks for your contribution! :tada: :+1:

Please assign me to review if it is ready!

**Comment by zwpaper** [comment]
Hi @gaocegege, this PR is currently done and could use nerdctl to build envs, I will raise another PRs for `up` envs and others.

but we still need to wait for buildkit to release a new version after this PR gets merged https://github.com/moby/buildkit/pull/3472, so that we can use a buildkitd with nerdctl support

**Comment by zwpaper** [comment]
Hi @gaocegege this should be ready for review.

before using nerdctl to build envs, please make sure:
1. nerdctl is inside your PATH, for lima case, please link nerdctl to nerdctl.lima
2. use buildkit image built from master, `zwpaper/buildkit:60e82c` if trust me and don't want to built it yourself

**Comment by gaocegege** [comment]
Thanks!

**Comment by gaocegege** [comment]
I think we could push the buildkit master image to tensorchord/buildkit, or use moby/buildkit:latest, WDYT?

**Review by gaocegege — COMMENTED** [review]
This PR is to add a new builder type nerd-container, right? Currently we have to use docker to run the image built in nerd buildkitd container.

/cc @kemingy @VoVAllen

**Comment by zwpaper** [comment]
there is an image `moby/buildkit:master`, it was built after I update the PR, and it should be ready to use now.

this PR adds a builder for nerdctl so that we can:

1. add a context option for nerdctl
2. start the buildkitd container by nerdctl
3. use nerdctl to send the build request to buildkitd and then build the env

**Comment by zwpaper** [comment]
for dev and review cases, we could use the master image. when releasing, we should wait for the versioned image for buildkit

File: pkg/builder/build.go
Comment (gaocegege) [inline pkg/builder/build.go:210]: What does this change mean?

Comment (zwpaper) [reply pkg/builder/build.go:210]: the `authprovider` is upgraded with the `buildkit` client library, the NewDockerAuthProvider param is updated from upstream so that I have to change the args

Comment (gaocegege) [reply pkg/builder/build.go:210]: Gotcha.

File: pkg/driver/nerdctl/nerdctl.go
Comment (kemingy) [inline pkg/driver/nerdctl/nerdctl.go:44]: Seems this will override the `err`. For non (mac + lima) user, they will get an `exec: "nerdctl.lima": executable file not found in $PATH`, but actually they need `nerdctl` instead of `nerdctl.lima`.

Comment (zwpaper) [reply pkg/driver/nerdctl/nerdctl.go]: Hi @kemingy, thanks for pointing this out, I have updated the error to `errors.New("can not found nerdctl(or nerdctl.lima for mac) in PATH")`

**Review by kemingy — APPROVED** [review]

**Comment by zwpaper** [comment]
@gaocegege need one more approval, what's your opinion?

**Pull request merged**
